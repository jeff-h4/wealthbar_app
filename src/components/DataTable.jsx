import React from 'react'
import PropTypes from 'prop-types';
import Table from 'react-bootstrap/Table';

export default class DataTable extends React.Component {
  constructor(props) {
    super(props);
    this.renderTableHeader = this.renderTableHeader.bind(this);
    this.renderRow         = this.renderRow.bind(this);
    this.formatTableData   = this.formatTableData.bind(this);
    this.extractValues     = this.extractValues.bind(this);
  }

  renderTableHeader() {
    return (
      <thead>
        <tr>
          <th></th>
          <th>TFSA</th>
          <th>RRSP</th>
        </tr>
      </thead>
    )
  }

  renderRow(key, rowData) {
    return (
      <tr>
        <th>{rowData.header}</th>
        {rowData.values.map((entry, index) => <td key={index}>{entry}</td>)}
      </tr>
    )
  }

  extractValues(data, targetValue) {
    return Object.keys(data).map(k => data[k][targetValue]);
  }

  formatTableData(inputData) {
    return {
      startingValue: {
        header: 'Initial Account Deposit',
        values: this.extractValues(inputData,'startingValue').map(x => x.toFixed(2))
      },
      futureValue: {
        header: 'Future Value',
        values: this.extractValues(inputData,'futureValue').map(x => x.toFixed(2))
      },
      taxOnWithdrawal: {
        header: 'Tax on Withdrawal',
        values: this.extractValues(inputData,'taxOnWithdrawal').map(x => x.toFixed(2))
      },
      afterTaxFutureValue: {
        header: 'After Tax Future Value',
        values: this.extractValues(inputData,'afterTaxFutureValue').map(x => x.toFixed(2))
      }
    }
  }

  render() {
    if (this.props.data === null) return null;

    let tableData = this.formatTableData(this.props.data);
    return (
      <Table striped>
        {this.renderTableHeader()}
        <tbody>
          {this.renderRow(1, tableData.startingValue)}
          {this.renderRow(2, tableData.futureValue)}
          {this.renderRow(3, tableData.taxOnWithdrawal)}
          {this.renderRow(4, tableData.afterTaxFutureValue)}
        </tbody>
      </Table>
    );
  }
}

DataTable.propTypes = {
  data: PropTypes.shape({
    tfsaResult: PropTypes.object,
    rrspResult: PropTypes.object
  })
}
