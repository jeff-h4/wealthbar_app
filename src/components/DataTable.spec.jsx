import React from 'react';
import renderer from 'react-test-renderer';

import DataTable from './DataTable';

describe('DataTable', () => {
  test('renders null if no data passed in', () => {
    const component = renderer.create(<DataTable data={null}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders datatable if data passed in', () => {
    let data = {
      tfsaResult: {startingValue: 100, futureValue: 200, taxOnWithdrawal: 0, afterTaxFutureValue: 200},
      rrspResult: {startingValue: 100, futureValue: 200, taxOnWithdrawal: 20, afterTaxFutureValue: 180}
    }
    const component = renderer.create(<DataTable data={data}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});