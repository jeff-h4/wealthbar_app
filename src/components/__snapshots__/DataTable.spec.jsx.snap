// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable renders datatable if data passed in 1`] = `
<table
  className="table table-striped"
>
  <thead>
    <tr>
      <th />
      <th>
        TFSA
      </th>
      <th>
        RRSP
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>
        Initial Account Deposit
      </th>
      <td>
        100.00
      </td>
      <td>
        100.00
      </td>
    </tr>
    <tr>
      <th>
        Future Value
      </th>
      <td>
        200.00
      </td>
      <td>
        200.00
      </td>
    </tr>
    <tr>
      <th>
        Tax on Withdrawal
      </th>
      <td>
        0.00
      </td>
      <td>
        20.00
      </td>
    </tr>
    <tr>
      <th>
        After Tax Future Value
      </th>
      <td>
        200.00
      </td>
      <td>
        180.00
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`DataTable renders null if no data passed in 1`] = `null`;
